experimental/juvix/contract.jvx

```diff
@@ -0,0 +1,198 @@
+    version = Nat.T;
+    name    = String.T;
+    symbol  = Char.T;
+    owner   = Address;
+  }
+
+
+Transaction =
+  type Transfer = {
+    from-account : Token.Address;
+    to-account   : Token.Address;
+    ammount      : Nat.T;
+  }
+
+  type Mint = {
+    mint-amount     = Nat.T;
+    mint-to-account = Token.Address;
+  }
+
+  type Burn = {
+    burn-amount       = Nat.T;
+    burn-from-account = Token.Address;
+  }
+
+  type Data =
+    | Transfer : Transfer -> Data
+    | Mint     : Mint     -> Data
+    | Burn     : Burn     -> Data
+
+  type T = {
+    data               : Data;
+    authorized-account : Token.Address;
+  }
+
+
+has-n : Accounts.T -> Token.Address -> Nat -> Bool
+has-n accounts add to-transfer =
+  case Accounts.select accounts add of
+  | Just n  -> to-transfer <= n
+  | Nothing -> False
+
+
+account-sub : acc : Accounts.T
+           -> add : Token.Address
+           -> num : Nat.T {has-n acc add num}
+           -> Accounts.T
+account-sub accounts add number =
+  case Accounts.select accounts add of
+  | Just balance ->
+     Accounts.put accounts ~key:add ~val:(balance - number)
+
+account-add : Accounts.T -> Token.Address -> Nat.T -> Accounts.T
+account-add accounts add number =
+  Accounts.update accounts ~f:(number +) ~key:add
+
+
+
+transfer-stor : stor   : Token.Storage
+              -> ~from : Token.Address
+              -> ~to   : Token.Address
+              -> num   : Nat.T {has-n stor.accounts from num}
+              -> Token.Storage
+transfer-stor stor ~from:add_from ~to:add_to num =
+  let new-acc = account-add (account-sub stor.accounts add_from) add-to num in
+  -- This is valid, because the monoid and measure of account track this
+  { total-supply = stor.total-supply
+  ; accounts     = new-acc
+  }
+
+
+
+Validation : Module
+  type T = Token.T -> Transaction.T -> Bool
+
+  transfer : T
+  burn     : T
+  mint     : T
+Validation =
+  type T = Token.T -> Transaction.T -> Bool
+
+  mint token tx =
+    case tx.data of
+    | Transaction.Mint ->
+      token.owner == tx-tx-authorized-account
+    | Transaction.Burn _ | Transaction.Transfer _ ->
+      false
+
+  transfer token tx =
+    case tx.data of
+    | Transaction.Transfer {from-account; amount} ->
+      has-n token.storage.accounts from-account amount
+      && tx.authroized-account == from-account
+    | Transaction.Burn _ | Transaction.Mint _ ->
+      false
+
+  Burn token tx =
+    case tx.data of
+    | Transaction.Burn {burn-from-account; burn-ammount} ->
+      has_n token.storage.accounts burn-from-account burn-amount
+      && tx.authroized-account == burn-from-account
+    | Transaction.Mint _ | Transaction.Transfer _ ->
+      false
+
+
+Operation =
+
+  type TokenTransaction f =
+    tok : Token.T -> tx : tx { f tok tx } -> Token.T
+
+  transfer : TokenTransaction Validation.transfer
+  transfer token transaction =
+    case transaction.data of
+    | Transfer {from-account; to-account; amount} ->
+      { token
```

Comment:
Is this the record update syntax? How would it work for updating more than one field?

Reply:
with another
`<field> = <expression>`

Reply:
Hmm, I wonder if there's a way to do it without a keyword / that would also work well inline.

Reply:
If there are two words in a row

```ocaml
{ token
  storage = 
}
```
we could probably assume record updating IFF, token is a record, though I'm not sure if this has any conflicts

Reply:
Right, I think that's probably too implicit though, `with` is fine for now.